models/classes/Security/Business/Service/SecretKeyService.php

```diff
@@ -0,0 +1,58 @@
+<?php declare(strict_types=1);
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA;
+ */
+
+namespace oat\taoLti\models\classes\Security\Business\Service;
+
+use Exception;
+use oat\tao\model\service\InjectionAwareService;
+use oat\taoLti\models\classes\Security\Business\Contract\SecretKeyServiceInterface;
+use oat\taoLti\models\classes\Security\Business\Domain\Exception\SecretKeyGenerationException;
+
+final class SecretKeyService extends InjectionAwareService implements SecretKeyServiceInterface
```

Comment:
Why is it final? What if for one customer we will need to have a specific implementation of such service and in a case when it is final we will be blocked. The solution will be in duplicating the service from scratch with duplication of the code instead of pure inheritance.

Reply:
Thank you for your feedback!
But I must disagree. The class is marked as `final` exactly to hint a usage of the interface, or composition instead of inheriting the concrete service. So no, there is a proper solution if you apply composition and delegation should you feel an urge of extending the service.